### PR TITLE
fix(typetests): Don't fail when packages have no dependencies

### PR DESCRIPTION
**Pull request by tylerbutler**
Type test preparation was failing when trying to update the devDependencies of a package without the devDependencies property in package.json. I fixed this by doing an undefined check.

I also tightened the typing of our PackageJson type by making dependencies and devDependencies nullable properties. This did require a few updates to handle the undefined checks, but overall that was a small price to pay for better type safety.

Finally, I added a catch handler to the code that executes the per-package function. This gives us an opportunity to log the package that's causing the error along with the stack trace. This will make future debugging easier.

### CODE REVIEW

**Review by jatgarg — APPROVED** [review]

**Review by alexvy86 — APPROVED** [review]

File: build-tools/packages/build-cli/src/commands/typetests.ts [hunk 1]
```diff
@@ -163,14 +163,20 @@ function updateTypeTestConfiguration(pkgJson: PackageJson, options: TypeTestConf
 
 		if (!enabled || options.version === VersionOptions.Clear) {
 			// There isn't a type safe alternative to delete here,
-			// and delete should be safe since the key can't collide with anything builtin since it has to end in `-previous` which makes it not a valid identifier.
-			// eslint-disable-next-line @typescript-eslint/no-dynamic-delete
-			delete pkgJson.devDependencies[oldDepName];
+			// and delete should be safe since the key can't collide with anything builtin since it has to end in `-previous`
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/commands/typetests.ts:166]: I think:
`pkgJson.devDependencies ??= {};`
is a bit more clear then adding these extra cheeks.

But what is more interesting here is that the type is wrong, so the compiler didn't catch it. The PackageJson type incorrectly assumes devDependencies exist and thats the real source of this bug in my opinion.

Either that type should be updated and/or when we convert the parsed data to that type we should check it (for example use a TypeBox schema to validate that it matches the type when doing the conversion so any future case where they mismatch error more clearly).

**Comment by CraigMacomber** [comment]
pkgJson.devDependencies is does not do any file system access: its just one blob of policy logic. That makes is very easy to unit test. I recommend adding a regression test for this case to the corresponding tests file. Maybe also add a test for the more common case as well since there isn't one.

File: build-tools/packages/build-cli/src/commands/typetests.ts (hunk 1 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/typetests.ts:166]: > But what is more interesting here is that the type is wrong, so the compiler didn't catch it. The PackageJson type incorrectly assumes devDependencies exist and thats the real source of this bug in my opinion.

Agreed. That types bug should be fixed in this PR as well.

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 2]
```diff
@@ -664,9 +664,12 @@ export async function setPackageDependencies(
 		if (dep !== undefined) {
 			const isSameReleaseGroup = MonoRepo.isSame(dep.pkg.monoRepo, pkg.monoRepo);
 			if (!isSameReleaseGroup || (updateWithinSameReleaseGroup && isSameReleaseGroup)) {
-				const dependencies = dev
+				let dependencies = dev
 					? pkg.packageJson.devDependencies
 					: pkg.packageJson.dependencies;
+				if (dependencies === undefined) {
+					dependencies = {};
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/lib/package.ts:671]: I don't think this works. If you do this, the line below which modifies dependencies modifies this new object, not the actual packageJson object.

It would be good to have a test for this showing that it works: that makes code review easier as we don't have to look for bugs which have test coverage.

Comment (CraigMacomber) [reply build-tools/packages/build-cli/src/lib/package.ts:671]: Note that testing this cleanly would likely require factoring out the updating logic like the typeTests code does.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 3]
```diff
@@ -2,11 +2,330 @@
...
+			});
+		});
+
+		describe("previousAndResetFalse", () => {
+			const name = "previousAndResetFalse";
+			const options = optionsMatrix.get(name)!;
+			it(`packageMinimal`, () => {
+				const pkg = clonedeep(packageMinimal);
+				updateTypeTestConfiguration(pkg, options);
+
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:160]:
```suggestion
				updateTypeTestConfiguration(pkg, options);
```

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 4]
```diff
@@ -30,4 +349,6 @@ describe("typetests tests", () => {
 			});
 		}
 	});
+
+	describe("previousVersion", () => {});
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:353]:
```suggestion
```

File: .vscode/settings.json [hunk 5]
```diff
@@ -76,7 +76,7 @@
 		"**/*.tsbuildinfo": true,
 		"**/_api-extractor-temp": true,
 		"**/dist/*": true,
-		"**/lib/*": true,
+		"{azure,common,docs,examples,experimental,packages,tools}/**/lib/*": true,
```
Comment (tylerbutler) [inline .vscode/settings.json:79]:
```suggestion
		"**/lib/*": true,
```

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 6]
```diff
@@ -2,11 +2,330 @@
...
+	scripts: {},
+};
+
+/**
+ * A test package.json with an enabled typeValidation node.
+ */
+const packageWithTypeValidation: PackageJson = {
+	name: "test-package",
+	version: "2.0.0-internal.6.0.0",
+	scripts: {},
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:31]:
```suggestion
	...packageMinimal,
```

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 7]
```diff
@@ -2,11 +2,330 @@
...
+	},
+};
+
+/**
+ * A test package.json with an disabled typeValidation node.
+ */
+const packageWithTypeValidationDisabled: PackageJson = {
+	name: "test-package",
+	version: "2.0.0-internal.6.0.0",
+	scripts: {},
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:52]:
```suggestion
	...packageMinimal,
```

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 8]
```diff
@@ -2,11 +2,330 @@
...
+			"broken-package": {
+				backCompat: false,
+				forwardCompat: false,
+			},
+		},
+		disabled: true,
+	},
+};
+
+const optionsMatrix: Map<string, TypeTestConfigActions> = new Map([
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:64]: You are taking a testing strategy here which assumes TypeTestConfigActions.version and TypeTestConfigActions.resetBroken need to be tested together.

If we look at the implementation (thus doing https://en.wikipedia.org/wiki/White-box_testing ), its clear that these two aspects are actually orthogonal and can be tested independently. This does two important things:
1. Allows us to have far fewer tests since we don't need to test combinations.
2. Allows our tests to be simpler since each one only needs to be concerned about one of these two aspects.

If we really wanted to, we could refactor the method we are testing into two separate methods, one for each aspect, but I don't think that is worth it (it would help prevent the tests from becoming out of date if the code was refactored to these two aspect were coupled, but such a change is unlikely, and would require tests updates anyway, and doing such a split just makes more methods/APIs to worry about).

Thus rather than resorting to table driven property testing here, I'd recommend a set of tests for setting the version, and a test for reset broken.

Comment (CraigMacomber) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:64]: Also this setup seems a bit odd to me. Normally you either:
1. define a list or matix of test cases like this, and loop over them generating one test for each which asserts that some properties hold
2. write a test for the different cases, each setting up its inputs and checking it specific requirments hold.

From what I can tell here you have combined these such that the data is seperated from the logic which processes it (case 1), but have seperate logic for each case. This defeats the usecase for case 1 (combining all the logic into one) and makes the tests hard to follow because the logic and case are seperated. This also makes it harder to specialize each test case for the specific things it needs to check. It also makes it easy to have bugs in the testing where cases here are accidently not used below.

This is such a supprising approach that when seeing this list of test cases, I assumed below it would be a loop over them asserting particularl properties hold.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 9]
```diff
@@ -2,11 +2,330 @@
...
+		];
+
+		describe("previousAndResetTrue", () => {
+			const name = "previousAndResetTrue";
+			const options = optionsMatrix.get(name)!;
+			it(`packageMinimal`, () => {
+				const pkg = clonedeep(packageMinimal);
+				updateTypeTestConfiguration(pkg, options);
+				expect(pkg.typeValidation).to.be.undefined;
+				expect(pkg.typeValidation?.broken).to.be.undefined;
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:124]: Why are you doing this? If we know pkg.typeValidation is undefined, pkg.typeValidation?.broken has to be undefined: thats how the `?.` operator works.
Same with the below line.

Adding usless asserts to tests distracts from what its actually testing for: we should aim for targeted tests where its very clear what is being tested and why. Otherwise its hard to tell if something has test coverage or not and you get more test code which is harder to maintain.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 10]
```diff
@@ -2,11 +2,330 @@
...
+				expect(pkg.typeValidation?.disabled).to.be.undefined;
+				expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:126]: no reason to check something exists before checking it has a specifc value. the quality check will handle undefiend just fine.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 11]
```diff
@@ -2,11 +2,330 @@
...
+				expect(pkg.typeValidation?.disabled).to.be.true;
+				expect(pkg.devDependencies?.["test-package-previous"]).not.to.exist;
+			});
+		});
+
+		describe("clearAndResetTrue", () => {
+			const name = "clearAndResetTrue";
+			const options = optionsMatrix.get(name)!;
+			it(`packageMinimal`, () => {
+				const pkg = clonedeep(packageMinimal);
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:197]: rathert than using a clone function, if you make `packageMinimal` into a factory function instead of a constant, you can get a fresh opbject each type, and have more confidence no test will accidently modify it.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 12]
```diff
@@ -2,11 +2,330 @@
...
+				updateTypeTestConfiguration(pkg, options);
+				expect(pkg.typeValidation).not.to.be.undefined;
+				expect(pkg.typeValidation?.broken).to.exist;
+				expect(pkg.typeValidation?.broken["broken-package"]).to.exist;
+				expect(pkg.typeValidation?.disabled).to.be.false;
+				expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
+				expect(pkg.devDependencies?.["test-package-previous"]).to.equal(
+					"npm:test-package@2.0.0-internal.4.0.0",
+				);
+			});
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:315]: When checking several aspects of some produced object in a test, its often cleaner to read, has nicer errors, and easier to maintain if you just do a assert.deepEqual to a literal instead of checking all the fields seperatly.

File: build-tools/packages/build-cli/src/lib/package.ts [hunk 13]
```diff
@@ -667,6 +668,10 @@ export async function setPackageDependencies(
 				const dependencies = dev
 					? pkg.packageJson.devDependencies
 					: pkg.packageJson.dependencies;
+				if (dependencies === undefined) {
+					log?.warning(`Package has no dependencies or devDependencies: ${pkg.name}`);
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/lib/package.ts:672]: message makes it wound like there are both no deps and no dev deps, which is not the case. Also its inclear that this should be an warning (whats wrong with having no deps?).

Also warnings are not great as they tend to accumulate then bother people later when they wertent the one that caused them: I'd prefer errors or just working properly in most cases.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts [hunk 14]
```diff
@@ -129,21 +129,20 @@ export abstract class PackageCommand<
...
-				this.flags.concurrency,
-				async (details: PackageWithKind) => {
-					started += 1;
+			await async.mapLimit(packages, this.flags.concurrency, async (pkg: PackageWithKind) => {
+				started += 1;
+				updateStatus();
+				try {
+					await this.processPackage(pkg, pkg.kind);
+					succeeded += 1;
+				} catch (error) {
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/BasePackageCommand.ts:138]: implicitly typing error as `any` is more unsafe than needed.


```suggestion
				} catch (error: unknown) {
```

is our preferred pattern.

I'd expect is to object to what you have here due to implicit any: do you not have noImplicit any enabled in this package?

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 10 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:126]: Won't there be better error messages though?

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 11 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:197]: Great idea.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 12 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:315]: Good idea.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 10 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:126]: I think using deepEqual will address my "better error messages" problem and is cleaner overall.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts (hunk 14 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/BasePackageCommand.ts:138]: Looks like that option is not enabled in our shared tsconfig: https://github.com/microsoft/FluidFramework/blob/c600fec19bd4f3d600293bb32d832d18218523d5/common/build/build-common/ts-common-config.json#L13

That seems... wrong. But also out of scope here. I'll set it to true for this project and see how many changes it requires.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 8 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:64]: > If we really wanted to, we could refactor the method we are testing into two separate methods, one for each aspect,
> but I don't think that is worth it (it would help prevent the tests from becoming out of date if the code was
> refactored to these two aspect were coupled

I know it's not strictly required, but I'm going to do it now in the spirit of making these tests and
accompanying code a "good example".

> This is such a supprising approach that when seeing this list of test cases, I assumed below it would be a loop over
> them asserting particularl properties hold.

I did loop originally, but ran into problems because some combinations needed different assertions. Splitting the
testing into independent pieces will address both of these problems and should make everything more readable.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 12 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts]: I ended up not using this in the simplified tests, but I'll remember this for next time.

File: build-tools/packages/build-cli/src/BasePackageCommand.ts (hunk 14 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/BasePackageCommand.ts]: I enabled noImplicitAny for the project, but it didn't appear to have any effect.

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/BasePackageCommand.ts]: > I enabled noImplicitAny for the project, but it didn't appear to have any effect.

Scratch that. It caused a build failure because of missing prettier types, which I've added. I wonder how hard it would be to enable this more broadly?

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 15]
```diff
@@ -2,11 +2,162 @@
...
+
+/**
+ * A test package.json with an enabled typeValidation node.
+ */
+function packageWithTypeValidation(): PackageJson {
+	return {
+		...packageMinimal(),
+		devDependencies: {
+			"test-package-previous": "npm:test-package@2.0.0-internal.4.0.0",
+			"another-dependency": "^1.0.0",
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:34]: rather than cluttering up every test here with this extra unrelated dependency, I think it might be better to have a single unit test that unrelated dependencies are left alone, then omit this from the rest of the tests.

Another similar way we can simplify most of the tests by splitting one case out into its own test is with versions.
We could make all the existing tests use a very simple version (even something as trivial as "1" should actually work fine). We can then test elsewhere that versions are handled in a compatible way that works with all the formats. Since this code uses "previousVersion" as to do all version related computations, such tests can be part of "previousVersion"'s test suite (which I think it already has) so this test suite can just use simpler versions.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 16]
```diff
@@ -2,11 +2,162 @@
...
+			it("removes previous test package dependency when type tests are disabled", () => {
+				const pkg = packageWithTypeValidationDisabled();
+				applyTypeTestVersionOptions(pkg, VersionOptions.Previous);
+				expect(pkg.devDependencies?.["test-package-previous"]).not.to.exist;
+			});
+		});
+	});
+
+	describe("resetBrokenTests", () => {
+		it(`ignores packages with no typeValidation node`, () => {
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:128]: I'm surprised we don't have linting which prevents use of `` formatting when "" would work. Its a little odd to be inconsistent with this, but it doesn't really matter.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 17]
```diff
@@ -2,11 +2,162 @@
...
+			const pkg = packageMinimal();
+			resetBrokenTests(pkg, true);
+			expect(pkg.typeValidation).to.not.exist;
+		});
+
+		it(`resets package with type tests`, () => {
+			const pkg = packageWithTypeValidation();
+			resetBrokenTests(pkg, true);
+			expect(pkg.typeValidation?.broken).to.be.empty;
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:137]: personally, I prefer `assert.deepEqual(pkg.typeValidation?.broken, [])`

Makes the empty case not special compared to other cases, and also is more obvious about what it does. For example as someone who doesn't use chai a lot, I'm not confident that the assert as you have it would error on undefined or empty objects, empty maps etc, while asserting equal to [] is very clear.

File: build-tools/packages/build-cli/src/commands/typetests.ts [hunk 18]
```diff
@@ -153,29 +158,62 @@ export function previousVersion(version: string): string {
...
+	options: TypeTestConfigActions,
+): void {
 	if (options.version !== undefined) {
-		const oldDepName = `${pkgJson.name}-previous`;
+		applyTypeTestVersionOptions(pkgJson, options.version);
+	}
 
-		// Packages can explicitly opt out of type tests by setting typeValidation.disabled to true.
-		const enabled = pkgJson.typeValidation?.disabled !== true;
+	if (options.resetBroken !== undefined) {
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/commands/typetests.ts:172]: since TypeTestConfigActions has the property that undefined members mean don't do anything, we could make resetBroken `resetBroken?: true` instead of `boolean`. This would eliminate one case from being possible to express, and thus replace a case we might want to test with one the compiler can remove for us.

It also makes the semantics of resetBrokenTests better align with its name, since it would then always reset them.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 16 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:128]: This is a byproduct of refactoring. I'll clean it up.

File: build-tools/packages/build-cli/src/commands/typetests.ts [hunk 19]
```diff
@@ -153,29 +158,62 @@ export function previousVersion(version: string): string {
...
 	}
+}
 
-	if (options.resetBroken === true && pkgJson.typeValidation !== undefined) {
+/**
+ * Removes any `typeValidation.broken` entries from package.json.
+ *
+ * @internal
+ */
+export function resetBrokenTests(pkgJson: PackageJson, resetBroken: boolean): void {
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/commands/typetests.ts:215]: if you want to make this easily testable, you could be more specific about what it takes in, for example After applying the above suggestion as well:
```suggestion
export function resetBrokenTests(pkgJson: { typeValidation?: ITypeValidationConfig}): void {
```

This way a test for this function could be very trivial. I think this would be enough of a test suite for such a simple function:
```
assert.deepEqual(resetBrokenTests({}, {});
assert.deepEqual(resetBrokenTests({ typeValidation: {broken: {}}}, {});
```

(I'd put both in one test for simplicity)

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 17 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:137]: `broken` is an object. Should it be `assert.deepEqual(pkg.typeValidation?.broken, {})`?

File: build-tools/packages/build-cli/src/commands/typetests.ts [hunk 20]
```diff
@@ -153,29 +158,62 @@ export function previousVersion(version: string): string {
...
+		pkgJson.devDependencies ??= {};
+		pkgJson.devDependencies[oldDepName] = `npm:${pkgJson.name}@${newVersion}`;
 	}
+}
 
-	if (options.resetBroken === true && pkgJson.typeValidation !== undefined) {
+/**
+ * Removes any `typeValidation.broken` entries from package.json.
+ *
+ * @internal
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/commands/typetests.ts:213]: You shouldn't mark package internal APIs with @internal. No release set at all means package internal, and @internal means public to other packages with the same maintainer: https://api-extractor.com/pages/tsdoc/tag_internal/

File: build-tools/packages/build-cli/src/commands/typetests.ts (hunk 19 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/typetests.ts:215]: This would require me to export the ITypeValidationConfig type from build-tools, which I'd really like to avoid. I guess I could copy the definition for ITypeValidationConfig, but that doesn't seem correct either.

File: build-tools/packages/build-cli/package.json [hunk 21]
```diff
@@ -121,6 +121,7 @@
 		"@types/inquirer": "^8.2.6",
 		"@types/mocha": "^9.1.1",
 		"@types/node": "^14.18.53",
+		"@types/prettier": "^2.7.3",
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/package.json:124]: What does this help with? Other packages don't seem to use it.

File: build-tools/packages/build-cli/src/commands/typetests.ts (hunk 18 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/typetests.ts:172]: This was new to me. In retrospect it's obvious but it never occurred to me to define a `true | undefined` type! One downside is that this usage doesn't work any more:

```ts
updateTypeTestConfiguration(json, { resetBroken: this.flags.reset, version });
```

I can massage the flag before calling the function, but is that better?

File: build-tools/packages/build-cli/package.json (hunk 21 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/package.json:124]: noImplicitAny requires it (I was wrong about no changes being needed). prettier is imported and called programmatically in the generate:changeset command.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 22]
```diff
@@ -2,11 +2,162 @@
...
+	};
+}
 
 describe("typetests tests", () => {
+	describe("applyTypeTestVersionOptions", () => {
+		describe("VersionOptions.Clear", () => {
+			it("removes previous test package dependency", () => {
+				const pkg = packageWithTypeValidation();
+				applyTypeTestVersionOptions(pkg, VersionOptions.Clear);
+				expect(pkg.devDependencies?.["test-package-previous"]).not.to.exist;
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:76]: when asserting that soemthing doesn't exist, its really easy to have a bug in your test where it never exists instead of was properly removed. (Imagine the case where you had a typo in "test-package-previous", either here or in the origional data: the test would pass even if applyTypeTestVersionOptions didn't work)

If you add an assert that it does exist before you try and remove it (or make this super clear by having the input data in the test dorectly), its a bit more robust.

Another way to reduce risk of this issue is asserting the parent object is equal to something. That way you don't have to explicitly name the thing you expect to be missing so you can't typo it. Ex:
`assert.deepEqual(pkg.devDependencies, {})` (This would work if you didn't have the extra dep in this tests input)

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 17 above)
Comment (CraigMacomber) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:137]: yes, and thats a nice reason to use deepEqual instead of to.be.empty: it makes it more clear what its doing :)

File: build-tools/packages/build-cli/src/commands/typetests.ts (hunk 18 above)
Comment (CraigMacomber) [reply build-tools/packages/build-cli/src/commands/typetests.ts:172]: With an optional boolean we have 4 ways the flag could be setup:
{x: true}
{x: false}
{x: undefined}
{}

With optional true, we have three:
{x: true}
{x: false}
{x: undefined}
{}

TypeScript sometimes sorta treats the last two differently (and JS clearly treats them differently). If we prefer the last one we can do the conversion inline as:
`{ ...(this.flags.reset? {resetBroken: true}, {}), version }`
If we prefer the second:
`{ resetBroken: this.flags.reset? true : undefined, version }`

A better alternative to this might be to just delete the updateTypeTestConfiguration function and the TypeTestConfigActions type and just repalce:

```
updateTypeTestConfiguration(json, { resetBroken: this.flags.reset, version });
```

with 

```typescript
	if (version !== undefined) {
		applyTypeTestVersionOptions(pkgJson, version);
	}

	if (this.flags.reset) {
		resetBrokenTests(pkgJson);
	}
```

We can just delete the whole mid level abstraction.
That moves slightly more logic into processPackage which is less testable, but deletes enough concepts that its likley a win.

If we ever to want to test all the actual flag interpertation logic, we could hoist the whole update json callback from processPackage out into a named function that takes the flags object and give it a basic integration test.

Comment (CraigMacomber) [reply build-tools/packages/build-cli/src/commands/typetests.ts:172]: Thats a specific application of the general approach of "if a data conversion seems like it would be nice, try and just delete enough abstractions/layers so the conversion isn't needed in the first place approach".

File: build-tools/packages/build-cli/src/commands/typetests.ts (hunk 19 above)
Comment (CraigMacomber) [reply build-tools/packages/build-cli/src/commands/typetests.ts:215]: ITypeValidationConfig is part of the already package exported PackageJson type. Exporting a type but not the types it referenced is questionable, and API-Extractor will reject it if you use our strict configuration.

Generally forcing someone to use a type made up of parts they can't refer to because they arn't exported is bad: it makes it hard to work with the type, hard to get docs for it etc. That is why API-Extractor consideres it an error.

Also the type is effectivly already exported as `PackageJson["typeValidation"]`, and has to appear in the package d.ts file since its used in that file: you are not losing any real encapsulation by exprting it with a proper name.

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/typetests.ts:215]: Good points.

File: build-tools/packages/build-cli/src/commands/typetests.ts (hunk 18 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/typetests.ts:172]: I removed the whole abstraction. Like you said, it wasn't adding anything.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 15 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:34]: > rather than cluttering up every test here with this extra unrelated dependency, I think it might be better to have a single unit test that unrelated dependencies are left alone, then omit this from the rest of the tests.

Good idea; implemented. I also used simpler versions since, as you pointed out, that logic is already tested independently.

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 22 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:76]: > If you add an assert that it does exist before you try and remove it (or make this super clear by having the input data in the test dorectly), its a bit more robust.

I really like the clarity this added. I _may_ have gone overboard, but in places where the input data is created elsewhere (like when using the factyory functions) checking before and after seems to add clarity and protect against test data changes that would silently pass.

> Another way to reduce risk of this issue is asserting the parent object is equal to something. That way you don't have to explicitly name the thing you expect to be missing so you can't typo it. Ex: assert.deepEqual(pkg.devDependencies, {}) (This would work if you didn't have the extra dep in this tests input)

I used this in a couple of places where it seemed clearer. But I didn't use it everywhere I could have because it seems misleading to assert the parent is empty. For example:

```ts
const pkg = packageWithTypeValidation(/* enabled */ false);
expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
updateTypeTestDependency(pkg, VersionOptions.ClearIfDisabled);
expect(pkg.devDependencies?.["test-package-previous"]).to.not.exist;
```

That seems clearer and more accurate to the intent of the test than:

```ts
const pkg = packageWithTypeValidation(/* enabled */ false);
expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
updateTypeTestDependency(pkg, VersionOptions.ClearIfDisabled);
expect(pkg.devDependencies).to.deep.equal({});
```

File: build-tools/packages/build-cli/src/commands/typetests.ts (hunk 19 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/typetests.ts]: I exported the typetest related types, but there are others that should be exported. I'll leave those for a future change.

File: build-tools/packages/build-cli/src/commands/typetests.ts [hunk 23]
```diff
@@ -101,27 +113,28 @@ If targeting prerelease versions, skipping versions, or using skipping some alte
...
-	 * If empty string, remove previous version.
-	 */
-	version?: string | VersionOptions;
-
-	/**
-	 * If true delete "broken" entries (therefore enabling the tests for them again).
-	 */
-	resetBroken?: boolean;
-}
+// /**
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/src/commands/typetests.ts:122]: Did you intend to delete these?

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/typetests.ts:122]: 🤦🏻 yeah

File: build-tools/packages/build-cli/test/commands/typetests.test.ts [hunk 24]
```diff
@@ -2,11 +2,173 @@
...
+			const pkg = packageWithTypeValidation();
+			resetBrokenTests(pkg);
+			expect(pkg.typeValidation?.broken).to.deep.equal({});
+		});
+
+		it("resets even if type validation is disabled", () => {
+			const pkg = packageWithTypeValidation(/* enabled */ false);
+			resetBrokenTests(pkg);
+			expect(pkg.typeValidation?.broken).to.deep.equal({});
+		});
```
Comment (CraigMacomber) [inline build-tools/packages/build-cli/test/commands/typetests.test.ts:169]: Tests have a cost (implemetaintion, maintaince, review, updating when the implementation changes etc). I feel like this test suite has a complexity much higher than the implementation.

One of the reasons I suggested making the input type so much more constrained is that now we can trust the compiler to ensure this function prabably doesn't screw up the fields it doesn't know about. This isn't guarenteed, but it should be enough for the level of confidence we want here.


Thus I think these last three tests really wouldn't be worth writing (they have more coupling than the other more targeted ones since they involve your full package information). I'd view these as more integration level tests, and with good unit testing and precise typing, we shouldn't need them. That said they seem reasonable and you already made them and I already reviewed them, so keeping them is also ok.

File: build-tools/packages/build-tools/src/index.ts [hunk 25]
```diff
@@ -5,11 +5,21 @@
 
 export { Context } from "./common/context";
 export { GitRepo } from "./common/gitRepo";
-export { FluidRepo, type VersionDetails } from "./common/fluidRepo";
+export {
+	type BrokenCompatTypes,
```
Comment (CraigMacomber) [inline build-tools/packages/build-tools/src/index.ts:9]: I generally don't see a reason to include `type` here. Unless there is both a type and non-type value, it does the same thing. I only use import/export type explicitly when I want to ensure that the non-type value (if there is one) isn't used, usually because using it would cause an initialization time cyclic dependency issue. Nothing wrong with including it though I just haven't see this done this way.

**Review by CraigMacomber — APPROVED** [review]

File: build-tools/packages/build-cli/test/commands/typetests.test.ts (hunk 24 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/test/commands/typetests.test.ts:169]: > I'd view these as more integration level tests, and with good unit testing and precise typing, we shouldn't need them.

Good point about integration vs. unit tests. I get nervous I'm not considering all the relevant cases, Need more experience.

> That said they seem reasonable and you already made them and I already reviewed them, so keeping them is also ok.

Thanks, I'll leave them and get this merged, but I think I get what you mean.

File: build-tools/packages/build-tools/src/index.ts (hunk 25 above)
Comment (tylerbutler) [reply build-tools/packages/build-tools/src/index.ts:9]: I like it because it's explicit about the intended use of the imported item. If it's only used as a type, import it as a type (IMO). It's also clearer that it will be completely removed from the JS and should have zero runtime impact. But I think if I really cared about that, I would push to enable [the lint rule](https://typescript-eslint.io/rules/consistent-type-imports/).
